fhir-server/src/main/java/au/csiro/pathling/aggregate/FreshAggregateExecutor.java

```diff
@@ -137,16 +138,16 @@ public AggregateResponse execute(@Nonnull final AggregateRequest query) {
   }
 
   @Nonnull
-  private Dataset<Row> joinGroupingsAndFilters(@Nonnull final Collection<FhirPath> groupings,
+  private Dataset<Row> joinGroupingsAndFilters(final FhirPath inputContext,
+      @Nonnull final Collection<FhirPath> groupings,
       @Nonnull final Collection<FhirPath> filters, @Nonnull final Column idColumn) {
     final List<Dataset<Row>> datasets = groupings.stream()
         .map(grouping -> {
           // We need to remove any trailing null values from non-empty collections, so that
           // aggregations do not count non-empty collections in the empty collection grouping. We do
           // this by joining the distinct set of resource IDs to the dataset using an outer join,
           // where the value is not null.
-          final Dataset<Row> distinctIds = grouping.getDataset().select(idColumn).distinct();
-          return join(grouping.getDataset(), idColumn, distinctIds,
+          return join(grouping.getDataset(), idColumn, inputContext.getDataset(),
```

Comment:
@piotrszul I made the change you suggested, and it works and fixes the problem of the duplicate evaluation of paths.